src/Fixer/ClassNotation/FinalInternalClassFixer.php

```diff
@@ -177,23 +214,62 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
             return false; // ignore class; it is abstract or already final
         }
 
-        $docToken = $tokens[$tokens->getPrevNonWhitespace($index)];
+        $decisions = [];
+        $currentIndex = $index;
 
-        if (!$docToken->isGivenKind(T_DOC_COMMENT)) {
-            return $this->configuration['consider_absent_docblock_as_internal_class'];
+        $acceptTypes = [
+            CT::T_ATTRIBUTE_CLOSE,
+            T_DOC_COMMENT,
+            T_COMMENT, // Skip comments
+        ];
+
+        if (\defined('T_READONLY')) {
+            // Skip readonly classes for PHP 8.2+
+            $acceptTypes[] = T_READONLY;
         }
 
-        $doc = new DocBlock($docToken->getContent());
+        while ($currentIndex) {
+            $currentIndex = $tokens->getPrevNonWhitespace($currentIndex);
+
+            if (!$tokens[$currentIndex]->isGivenKind($acceptTypes)) {
+                break;
+            }
+
+            if ($this->checkAttributes && $tokens[$currentIndex]->isGivenKind(CT::T_ATTRIBUTE_CLOSE)) {
+                $attributeStartIndex = $tokens->findBlockStart(Tokens::BLOCK_TYPE_ATTRIBUTE, $currentIndex);
+                $decisions[] = $this->isClassCandidateBasedOnAttribute($tokens, $attributeStartIndex, $currentIndex);
+
+                $currentIndex = $attributeStartIndex;
+            }
+
+            if ($tokens[$currentIndex]->isGivenKind([T_DOC_COMMENT])) {
+                $decisions[] = $this->isClassCandidateBasedOnPhpDoc($tokens, $currentIndex);
+            }
+        }
+
+        if (false !== array_search(false, $decisions, true)) {
+            return false;
+        }
+
+        if (false !== array_search(true, $decisions, true)) {
+            return true;
+        }
+
+        return !\count($decisions) && $this->configuration['consider_absent_docblock_as_internal_class'];
+    }
+
+    private function isClassCandidateBasedOnPhpDoc(Tokens $tokens, int $index): ?bool
+    {
+        $doc = new DocBlock($tokens[$index]->getContent());
         $tags = [];
 
         foreach ($doc->getAnnotations() as $annotation) {
             if (1 !== Preg::match('/@\S+(?=\s|$)/', $annotation->getContent(), $matches)) {
                 continue;
             }
-
             $tag = strtolower(substr(array_shift($matches), 1));
 
-            foreach ($this->configuration['annotation_exclude'] as $tagStart => $true) {
+            foreach ($this->configuration['exclude'] as $tagStart => $true) {
                 if (str_starts_with($tag, $tagStart)) {
```

Comment:
I know this is not your code, but this is wrong and maybe you want fix it while working on this? Consider class:

```php
/**
 * @foo
 * @abcdef
 */
final class A{}
```

With config `['include' => ['@foo'], 'exclude' => ['@abc']]` it won't mark class as `final`, even if it should because there's configured `include`, while `exclude` _should_ not match.

Reply:
Fixed it :)